crates/sui-adapter/src/programmable_transactions/linkage_view.rs

```diff
@@ -0,0 +1,258 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    cell::RefCell,
+    collections::{hash_map::Entry, BTreeMap, HashMap, HashSet},
+    str::FromStr,
+};
+
+use move_core_types::{
+    account_address::AccountAddress,
+    identifier::{IdentStr, Identifier},
+    language_storage::{ModuleId, StructTag},
+    resolver::{LinkageResolver, ModuleResolver, ResourceResolver},
+};
+use sui_types::{
+    base_types::ObjectID,
+    error::{ExecutionError, SuiError},
+    move_package::{MovePackage, TypeOrigin, UpgradeInfo},
+};
+
+use super::types::StorageView;
+
+/// Exposes module and linkage resolution to the Move runtime.  The first by delegating to
+/// `StorageView` and the second via linkage information that is loaded from a move package.
+pub struct LinkageView<'state, S: StorageView> {
+    /// Immutable access to the store for the transaction.
+    state_view: &'state S,
+    /// An explicit linkage context. (If none is set, then re-linking is disabled everywhere).
+    linkage_info: Option<LinkageInfo>,
+    /// Cache containing the type origin information from every package that has been set as the
+    /// link context, and every other type that has been requested by the loader in this session.
+    /// It's okay to retain entries in this cache between different link contexts because a type's
+    /// Runtime ID and Defining ID are invariant between across link contexts.
```

Comment:
Just to make sure - it means that for a given type, a query for the type origin will always be for the same `ModuleId` (including its address component), regardless of the current linkage context set, right?

Reply:
That's right -- both the queried ID (which is the "runtime", or "original" ID for the package that the type comes from) and the returned ID (which is the ID of the package that first introduced the type) will always remain the same.